src/validation.cpp

```diff
@@ -350,21 +350,29 @@ void CChainState::MaybeUpdateMempoolForReorg(
     // the disconnectpool that were added back and cleans up the mempool state.
     m_mempool->UpdateTransactionsFromBlock(vHashUpdate);
 
-    const auto check_final_and_mature = [this, flags=STANDARD_LOCKTIME_VERIFY_FLAGS](CTxMemPool::txiter it)
+    // Check whether the transaction is still final and, if it spends a coinbase output, mature.
+    // Update the entry's cached LockPoints if needed. If this returns true, the tx would be invalid
+    // in the next block, so the mempool must remove this entry and all of its descendants.
+    const auto filter_final_and_mature = [this, flags=STANDARD_LOCKTIME_VERIFY_FLAGS](CTxMemPool::txiter it)
         EXCLUSIVE_LOCKS_REQUIRED(m_mempool->cs, ::cs_main) {
-        bool should_remove = false;
         AssertLockHeld(m_mempool->cs);
         AssertLockHeld(::cs_main);
         const CTransaction& tx = it->GetTx();
         LockPoints lp = it->GetLockPoints();
         const bool validLP{TestLockPointValidity(m_chain, lp)};
         CCoinsViewMemPool view_mempool(&CoinsTip(), *m_mempool);
-        if (!CheckFinalTx(m_chain.Tip(), tx, flags)
-            || !CheckSequenceLocks(m_chain.Tip(), view_mempool, tx, flags, &lp, validLP)) {
-            // Note if CheckSequenceLocks fails the LockPoints may still be invalid
-            // So it's critical that we remove the tx and not depend on the LockPoints.
-            should_remove = true;
-        } else if (it->GetSpendsCoinbase()) {
+        // The transaction must be final.
+        if (!CheckFinalTx(m_chain.Tip(), tx, flags)) return true;
+        // Note if CheckSequenceLocks fails the LockPoints may still be invalid
+        // So it's critical that we remove the tx and not depend on the LockPoints.
+        if (!CheckSequenceLocks(m_chain.Tip(), view_mempool, tx, flags, &lp, validLP)) {
+            return true;
+        } else if (!validLP) {
```

Comment:
nit: `else` is not necessary after `return`. May as well remove it also for consistency with the above `if (!CheckFinalTx...` (which does not include `else` after `return`).

```suggestion
        }
        if (!validLP) {
```

Reply:
The `!validLP` check here is logically associated with the `!CheckSequenceLocks()` check, so it makes sense to have it in an `else` branch. Logically the two things are the same since the if branch contains an early return, but structuring it as if/else signals to the developer that the two things are connected.